src/OpenSpaceToolkit/Astrodynamics/NumericalSolver.cpp

```diff
@@ -373,6 +392,131 @@ Array<NumericalSolver::Solution> NumericalSolver::integrateDuration(
+
+    // initialize stepper
+    double currentTime = 0.0;
+    stepper.initialize(aStateVector, currentTime, signedTimeStep);
+
+    // do first step
+    double previousTime;
+    std::tie(previousTime, currentTime) = stepper.do_step(aSystemOfEquations);
+    observeNumericalIntegration(stepper.current_state(), stepper.current_time());
+
+    bool conditionSatisfied = false;
+    Real currentValue = Real::Undefined();
+
+    const Size stateSize = anInitialStateVector.size();
+
+    Real previousValue = anEventCondition->evaluate(
+        Eigen::Map<const Eigen::VectorXd>(stepper.current_state().data(), stateSize), stepper.current_time()
+    );
+    NumericalSolver::StateVector currentState;
+
+    // account for integration direction
+    const auto checkTimeLimit = [&currentTime, &aDurationInSeconds]()
+    {
+        if (aDurationInSeconds > 0.0)
+        {
+            return currentTime < aDurationInSeconds;
+        }
+        return currentTime > aDurationInSeconds;
+    };
+
+    while (checkTimeLimit())
+    {
+        std::tie(previousTime, currentTime) = stepper.do_step(aSystemOfEquations);
+        currentState = stepper.current_state();
+
+        currentValue =
+            anEventCondition->evaluate(Eigen::Map<const Eigen::VectorXd>(currentState.data(), stateSize), currentTime);
+
+        conditionSatisfied = anEventCondition->isSatisfied(currentValue, previousValue);
+
+        if (conditionSatisfied)
+        {
+            break;
+        }
+
+        observeNumericalIntegration(currentState, currentTime);
+
+        previousValue = currentValue;
+    }
+
+    // TBI: Share information upstream on if the condition was satisfied
+    if (!conditionSatisfied)
+    {
+        stepper.calc_state(aDurationInSeconds, currentState);
+        return {currentState, aDurationInSeconds};
+    }
+
+    // Search for the exact time of the condition change
+    NumericalSolver::StateVector midState(currentState);
+    double midTime;
+
+    // TBI: Make this a parameter
+    const Index maxIterationCount = 100;
+
+    for (Index iterationCount = 0; iterationCount < maxIterationCount; ++iterationCount)
+    {
+        midTime = 0.5 * (previousTime + currentTime);
+        stepper.calc_state(midTime, midState);
+
+        const Real midValue =
+            anEventCondition->evaluate(Eigen::Map<const Eigen::VectorXd>(midState.data(), stateSize), midTime);
+
+        if (anEventCondition->isSatisfied(midValue, previousValue))
+        {
+            // root lies between previousTime and midTime
+            // update current -> mid
+            currentTime = midTime;
+            currentValue = midValue;
+        }
+        else
+        {
+            // root lies between midTime and currentTime
+            // update previous -> mid
+            previousTime = midTime;
+            previousValue = midValue;
+        }
+
+        // TBI: Make tolerance a parameter
```

Comment:
```suggestion
        // TBI: Make tolerance a parameter
```

Neat, should it be added to `EventCondition` derived classes, yeah?

Reply:
actually, maybe even base class?